packages/vscode-dependencies-validation/test/diagnostics.spec.ts

```diff
@@ -56,24 +53,24 @@ describe("diagnostics unit test", () => {
     });
 
     it("there are 2 dependency issues", async () => {
+      const uri = <Uri>{ fsPath: packageJsonPath };
```

Comment:
these tests are understandable but could be better if the code they are testing was refactored.

```typescript
export async function refreshDiagnostics(
  uri: Uri,
  dependencyIssueDiagnostics: DiagnosticCollection
): Promise<void> {
  if (isNotInNodeModules(uri)) {
    const { problems } = await findDependencyIssues(uri.fsPath);

    const diagnostics = isEmpty(problems)
      ? []
      : [constructDiagnostic(problems)];

    dependencyIssueDiagnostics.set(uri, diagnostics);
  }
}
```

1. the filtering of `isNotInNodeModules` could be extracted outside this function
  - generally could we ever trigger a check for anything that is **inside** node_modules? is this filter really needed?
  - but lets assume it is needed
2. `findDependencyIssues` could be defined as a DI parameter so the deps of this function are more clear.
3. `dependencyIssueDiagnostics` could be defined as a **sub-set** of `DiagnosticCollection` APIs

Then we will have something like:

```typescript
// TODO; consider config object instead of three params
export async function refreshDiagnostics(
  uri: Uri,
  dependencyIssueDiagnostics: Pick<DiagnosticCollection, "set">,
  findDependencyIssues: async (path:string) =>  {problems:string[] }
): Promise<void> {
    // more explicit early exit.
    if (isInsideNodeModules(uri)) { 
        return;
    }

    const { problems } = await findDependencyIssues(uri.fsPath);
    const diagnostics = isEmpty(problems)
      ? []
      : [constructDiagnostic(problems)];
    dependencyIssueDiagnostics.set(uri, diagnostics);
}
```

Now this is easier to test without needing **proxyQuire** for anything but `vscode`.
The problem is that **proxyQuire** may need to be updated for any new import at the root of module.
- e.g if that new import transitivily needs `vscode` 
- or if the **unit** we are testing is using that new import.

And we won't automagically know if a test need to be updated due to production code changes (adding new import at the module level). But if a **unit** (function) needs another "DI" param, any test that calls it directly would fail **during TypeScript compilation**.

The risk of course is that we would need to pass too many of these DI params between functions
- but that is not the case here.
- This raises the question of evaluating using  proper DI framework for writing our VSCode exts (out of scope...)

Now you could write a simpler test:

```typescript

let refreshDiagnosticsProxy: any // todo: missing right type

beforeEach(() => {
      const diagnosticsModule = proxyquire("../src/diagnostics", {
        // Note: only proxying vscode
        vscode: {
          Range,
          Diagnostic,
          "@noCallThru": true,
        },
      });
      refreshDiagnosticsProxy = diagnosticsModule.refreshDiagnostics;
})

it("releant path outside node_modules", async () => {
      // note: preparations of the params are fully isolated in this code block
     //           without any external helpers / proxies.
      const uri = <Uri>{ fsPath: "/user/home/projects/app-studio-toolkit };
      const issues = ["missing: json-fixer@1.6.12", "missing: sometingelse@1.2.3];
      const findDependencyIssues: async () => isssues
      const diagnosticCollection = new Set<Uri, string[]>()

      // note the `Set` should match the DiagonsticColletion set method
      await refreshDiagnosticsProxy(uri, diagnosticCollectionMock, findDependencyIssues);

      // Note: we can assert on the Set object as it was modified, no need for Sinon here...
      const diagnosticsEntryResult = diagnosticCollection.get(uri);
      expect(diagnosticsEntryResult).to.have.lengthOf(2);
      expect(diagnosticsEntryResult).to.deep.eql(issues)
});

// now we test **inside** node modules
it("releant path inside node_modules", async () => {
      const uri = <Uri>{ fsPath: "/user/home/projects/app-studio-toolkit };
      const findDependencyIssues: async () => []
      const diagnosticCollection = new Set<Uri, string[]>()

      await refreshDiagnosticsProxy(uri, diagnosticCollectionMock, findDependencyIssues);
      expect(diagnosticCollection.has(uri)).to.be.false;
});
```

WDYT?

Reply:
I understand what you tried to achieve but cannot agree with this approach.
In this code you replace actual diagnosticCollection by new Set<Uri, string[]>().
Why do you believe that real DiagnosticCollection is based on Set ?
It is not more a unit test of DiagnosticCollection, it tests only Set